### PR TITLE
[spirv] Support interpolation modes on PS input variables

### PR DESCRIPTION
**Pull request by antiagainst**

### CODE REVIEW

File: tools/clang/lib/SPIRV/DeclResultIdMapper.cpp [hunk 1]
```diff
@@ -356,6 +356,39 @@ bool DeclResultIdMapper::createStageVars(const DeclaratorDecl *decl,
...
+      const QualType elemType = typeTranslator.getElementType(type);
+
+      if (elemType->isBooleanType() || elemType->isIntegerType()) {
+        // TODO: Probably we can call hlsl::ValidateSignatureElement() for the
+        // following check.
+        if (decl->getAttr<HLSLLinearAttr>() ||
+            decl->getAttr<HLSLCentroidAttr>() ||
+            decl->getAttr<HLSLNoPerspectiveAttr>() ||
+            decl->getAttr<HLSLSampleAttr>()) {
+          emitError("only nointerpolation mode allowed for integer input "
```
Comment (ehsannas) [inline tools/clang/lib/SPIRV/DeclResultIdMapper.cpp]: there's a discrepancy between the error message and the if condition.
it says "for integer input parameters", but the condition also checks for booleans.

File: tools/clang/lib/SPIRV/ModuleBuilder.cpp [hunk 2]
```diff
@@ -368,6 +368,29 @@ void ModuleBuilder::decorateLocation(uint32_t targetId, uint32_t location) {
...
+  case spv::Decoration::Flat:
+    d = Decoration::getFlat(theContext);
+    break;
+  case spv::Decoration::NoPerspective:
+    d = Decoration::getNoPerspective(theContext);
+    break;
+  case spv::Decoration::Sample:
+    d = Decoration::getSample(theContext);
+    break;
+  default:
```
Comment (ehsannas) [inline tools/clang/lib/SPIRV/ModuleBuilder.cpp]: no need for the default statement here...

File: tools/clang/lib/SPIRV/TypeTranslator.cpp [hunk 3]
```diff
@@ -278,6 +278,13 @@ bool TypeTranslator::isSpirvAcceptableMatrixType(QualType type) {
   return isMxNMatrix(type, &elemType) && elemType->isFloatingType();
 }
 
+QualType TypeTranslator::getElementType(QualType type) {
+  QualType elemType = {};
+  (void)(isScalarType(type, &elemType) || isVectorType(type, &elemType) ||
```
Comment (ehsannas) [inline tools/clang/lib/SPIRV/TypeTranslator.cpp]: just curious: is `(void)` necessary here?

File: tools/clang/lib/SPIRV/TypeTranslator.h [hunk 4]
```diff
@@ -81,6 +81,10 @@ class TypeTranslator {
   /// counts.
   static bool isSpirvAcceptableMatrixType(QualType type);
 
+  /// \brief Returns the <result-id> of the element type for the given
```
Comment (ehsannas) [inline tools/clang/lib/SPIRV/TypeTranslator.h]: I expect result-id to be `uint32_t`, not `QualType`
out of date comment?

File: tools/clang/test/CodeGenSPIRV/spirv.interpolation.hlsl [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+                       int    int_a: INTA;
+// CHECK: OpDecorate %in_var_INTD Flat
+  nointerpolation      int3   int_d: INTD;
+
+// CHECK: OpDecorate %in_var_UINTA Flat
+                       uint  uint_a: UINTA;
+// CHECK: OpDecorate %in_var_UINTD Flat
+  nointerpolation      uint3 uint_d: UINTD;
+
+// CHECK: OpDecorate %in_var_BOOLA Flat
```
Comment (ehsannas) [inline tools/clang/test/CodeGenSPIRV/spirv.interpolation.hlsl]: should this be applied even if `nointerpolation` is not explicitly added?

File: tools/clang/lib/SPIRV/DeclResultIdMapper.cpp (hunk 1 above)
Comment (antiagainst) [inline tools/clang/lib/SPIRV/DeclResultIdMapper.cpp]: Basically copied the logic from https://github.com/Microsoft/DirectXShaderCompiler/blob/55667043c389dcba990c678390abc9e7cc9fe1d3/lib/HLSL/DxilValidation.cpp#L3197. They checked both booleans and integers but report error message for only integers. I was trying to keep consistency. 

Besides, booleans are most of the time treated as integers. And I assume if `elemType->isBooleanType()` is true, `elemType->isIntergerType()` will also return true. The above check may be redundant, but it's clear.

File: tools/clang/lib/SPIRV/TypeTranslator.cpp (hunk 3 above)
Comment (antiagainst) [inline tools/clang/lib/SPIRV/TypeTranslator.cpp]: We are not using the result of this expression. Compilers may generate warnings with some switch turned on (-W unused variable or something). This is to suppress it and also make it clear we are not interested in the result.

File: tools/clang/test/CodeGenSPIRV/spirv.interpolation.hlsl (hunk 5 above)
Comment (antiagainst) [inline tools/clang/test/CodeGenSPIRV/spirv.interpolation.hlsl]: Yes. That's the default and only one supported for integers.

https://msdn.microsoft.com/en-us/library/windows/desktop/bb509668(v=vs.85).aspx

File: tools/clang/lib/SPIRV/ModuleBuilder.cpp (hunk 2 above)
Comment (antiagainst) [inline tools/clang/lib/SPIRV/ModuleBuilder.cpp]: Done

File: tools/clang/lib/SPIRV/TypeTranslator.h (hunk 4 above)
Comment (antiagainst) [inline tools/clang/lib/SPIRV/TypeTranslator.h]: Good catch. Updated.

**Comment by ehsannas** [comment]
Thanks for clarifications. 👍

**Pull request merged**